src/main.rs

```diff
@@ -63,1116 +66,895 @@ fn main() -> Result<()> {
+            let namespace_id = if let Some(binding) = namespace.binding {
+                commands::kv::get_namespace_id(&target, &binding)?
+            } else {
+                namespace
+                    .namespace_id
+                    .expect("Namespace ID is required if binding isn't supplied")
+            };
+
+            match subcommand {
+                KvBulk::Put { path } => {
+                    commands::kv::bulk::put(&target, &user, &namespace_id, &path)?
+                }
+                KvBulk::Delete { path } => {
+                    commands::kv::bulk::delete(&target, &user, &namespace_id, &path)?
+                }
             }
-            _ => unreachable!(),
         }
-    } else if let Some(matches) = matches.subcommand_matches("tail") {
-        let config_path = Path::new(
-            matches
-                .value_of("config")
-                .unwrap_or(commands::DEFAULT_CONFIG_PATH),
-        );
-        let manifest = settings::toml::Manifest::new(config_path)?;
-        let format = matches.value_of("format").unwrap().to_owned();
-        let env = matches.value_of("env");
-        let target = manifest.get_target(env, is_preview)?;
-        let user = settings::global_user::GlobalUser::new()?;
-
-        let tunnel_port: Option<u16> = matches
-            .value_of("tunnel_port")
-            .map(|p| p.parse().expect("--port expects a number"));
-        let metrics_port: Option<u16> = matches
-            .value_of("metrics_port")
-            .map(|p| p.parse().expect("--metrics expects a number"));
-
-        let verbose = matches.is_present("verbose");
-
-        commands::tail::start(&target, &user, format, tunnel_port, metrics_port, verbose)?;
-    } else if matches.subcommand_matches("login").is_some() {
-        commands::login::run()?;
-    } else if let Some(matches) = matches.subcommand_matches("report") {
-        commands::report::run(matches.value_of("log")).map(|_| {
-            eprintln!("Report submission sucessful. Thank you!");
-        })?;
+        Command::Tail {
+            format,
+            tunnel_port,
+            metrics_port,
+        } => {
+            let manifest = Manifest::new(&cli.config)?;
+            let target = manifest.get_target(cli.environment.as_deref(), false)?;
+            let user = settings::global_user::GlobalUser::new()?;
+
+            commands::tail::start(
+                &target,
+                &user,
+                format,
+                tunnel_port,
+                metrics_port,
+                cli.verbose,
+            )?;
+        }
+        Command::Login => {
+            commands::login::run()?;
+        }
+        Command::Report { log } => {
+            commands::report::run(log.as_deref()).map(|_| {
+                eprintln!("Report submission sucessful. Thank you!");
+            })?;
+        }
     }
+
     Ok(())
 }
+
+#[cfg(test)]
+mod tests {
```

Comment:
Similarly, these should probably all go in a separate module along with the AdhocMigration impl and other non-CLI things. What do you think?